app/services/metrics/base_metric_processor.rb

```diff
@@ -0,0 +1,84 @@
+module Metrics
+  ##
+  # Base class for classes that do the processing of a Metric from a
+  # collection of events.
+  class BaseMetricProcessor < BaseService
+    def initialize(events:, time_interval:)
+      @events = events
+      @time_interval = time_interval
+    end
+
+    ##
+    # Processes the given events to generate the metrics.
+    # The given @events are expected to include all the events during the full
+    # time interval ( [@starting_at, @starting_at + @time_span) ).
+    def call
+      process_events(events: @events, time_interval: @time_interval)
+    end
+
+    private
+
+    ##
+    # Processes the given events to generate the review_turnaround metrics.
+    def process_events(events:, time_interval:)
+      accumulators = create_accumulators
+
+      iterate_events(events: events, time_interval: time_interval, accumulators: accumulators)
+
+      update_metrics(time_interval: time_interval, accumulators: accumulators)
+    end
+
+    ##
+    # Processes all the given events that are included in the time interval
+    def iterate_events(events:, time_interval:, accumulators:)
+      events.each do |event|
```

Comment:
use `find_each` to prevent memory bloats https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_each